File: base-v3/src/main/java/com/adyen/checkout/base/model/payments/response/Threeds2Action.java
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 9/12/2020.
 */

package com.adyen.checkout.base.model.payments.response;

import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.adyen.checkout.base.util.ActionTypes;
import com.adyen.checkout.core.exception.ModelSerializationException;
import com.adyen.checkout.core.exception.NoConstructorException;
import com.adyen.checkout.core.model.JsonUtils;

import org.json.JSONException;
import org.json.JSONObject;

@SuppressWarnings("MemberName")
public class Threeds2Action extends Action {
    @NonNull
    public static final Creator<Threeds2Action> CREATOR = new Creator<>(Threeds2Action.class);

    public static final String ACTION_TYPE = ActionTypes.THREEDS2;

    private static final String TOKEN = "token";
    private static final String SUBTYPE = "subtype";

    @NonNull
    public static final Serializer<Threeds2Action> SERIALIZER = new Serializer<Threeds2Action>() {

        @NonNull
        @Override
        public JSONObject serialize(@NonNull Threeds2Action modelObject) {
            final JSONObject jsonObject = new JSONObject();
            try {
                // Get parameters from parent class
                jsonObject.putOpt(Action.TYPE, modelObject.getType());
                jsonObject.putOpt(Action.PAYMENT_DATA, modelObject.getPaymentData());
                jsonObject.putOpt(Action.PAYMENT_METHOD_TYPE, modelObject.getPaymentMethodType());

                jsonObject.putOpt(TOKEN, modelObject.getToken());
                jsonObject.putOpt(SUBTYPE, modelObject.getSubtype());
            } catch (JSONException e) {
                throw new ModelSerializationException(Threeds2Action.class, e);
            }
            return jsonObject;
        }

        @NonNull
        @Override
        public Threeds2Action deserialize(@NonNull JSONObject jsonObject) {
            final Threeds2Action threeds2ChallengeAction = new Threeds2Action();

            // getting parameters from parent class
            threeds2ChallengeAction.setType(jsonObject.optString(Action.TYPE, null));
            threeds2ChallengeAction.setPaymentData(jsonObject.optString(Action.PAYMENT_DATA, null));
            threeds2ChallengeAction.setPaymentMethodType(jsonObject.optString(Action.PAYMENT_METHOD_TYPE, null));

            threeds2ChallengeAction.setToken(jsonObject.optString(TOKEN, null));
            threeds2ChallengeAction.setSubtype(jsonObject.optString(SUBTYPE, null));
            return threeds2ChallengeAction;
        }
    };

    private String token;
    private String subtype;

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this));
    }

    @Nullable
    public String getToken() {
        return token;
    }

    public void setToken(@Nullable String token) {
        this.token = token;
    }

    @Nullable
    public String getSubtype() {
        return subtype;
    }

    public void setSubtype(@Nullable String subtype) {
        this.subtype = subtype;
    }

    @Nullable
    public SubType getSubtypeEnum() {
        return parseStringToSubType(subtype);
    }

    @Nullable
    private SubType parseStringToSubType(@Nullable String type) {
        if (type == null) {
            return null;
        }
        switch (type) {
            case SubTypes.FINGERPRINT:
                return SubType.FINGERPRINT;
            case SubTypes.CHALLENGE:
                return SubType.CHALLENGE;
            default:
                return null;
        }
    }

    private static final class SubTypes {

        public static final String FINGERPRINT = "fingerprint";
        public static final String CHALLENGE = "challenge";

        private SubTypes() {
            throw new NoConstructorException();
        }
    }

    public enum SubType {
        FINGERPRINT,
        CHALLENGE
    }
}

File: components-core/src/main/java/com/adyen/checkout/components/model/payments/response/Threeds2Action.kt
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by josephj on 9/12/2020.
 */
package com.adyen.checkout.components.model.payments.response

import android.os.Parcel
import android.os.Parcelable
import com.adyen.checkout.components.util.ActionTypes
import com.adyen.checkout.core.exception.ModelSerializationException
import com.adyen.checkout.core.model.JsonUtils
import com.adyen.checkout.core.model.getStringOrNull
import org.json.JSONException
import org.json.JSONObject

class Threeds2Action(
    override var type: String? = null,
    override var paymentData: String? = null,
    override var paymentMethodType: String? = null,
    var token: String? = null,
    var subtype: String? = null,
    var authorisationToken: String? = null
) : BaseThreeds2Action() {

    override fun writeToParcel(dest: Parcel, flags: Int) {
        JsonUtils.writeToParcel(dest, SERIALIZER.serialize(this))
    }

    companion object {
        const val ACTION_TYPE = ActionTypes.THREEDS2

        private const val TOKEN = "token"
        private const val SUBTYPE = "subtype"
        private const val AUTHORISATION_TOKEN = "authorisationToken"

        @JvmField
        val CREATOR: Parcelable.Creator<Threeds2Action> = Creator(Threeds2Action::class.java)

        @JvmField
        val SERIALIZER: Serializer<Threeds2Action> = object : Serializer<Threeds2Action> {
            override fun serialize(modelObject: Threeds2Action): JSONObject {
                return try {
                    JSONObject().apply {
                        putOpt(TYPE, modelObject.type)
                        putOpt(PAYMENT_DATA, modelObject.paymentData)
                        putOpt(PAYMENT_METHOD_TYPE, modelObject.paymentMethodType)
                        putOpt(TOKEN, modelObject.token)
                        putOpt(SUBTYPE, modelObject.subtype)
                        putOpt(AUTHORISATION_TOKEN, modelObject.authorisationToken)
                    }
                } catch (e: JSONException) {
                    throw ModelSerializationException(Threeds2Action::class.java, e)
                }
            }

            override fun deserialize(jsonObject: JSONObject): Threeds2Action {
                return try {
                    Threeds2Action(
                        token = jsonObject.getStringOrNull(TOKEN),
                        subtype = jsonObject.getStringOrNull(SUBTYPE),
                        authorisationToken = jsonObject.getStringOrNull(AUTHORISATION_TOKEN),
                        type = jsonObject.getStringOrNull(TYPE),
                        paymentData = jsonObject.getStringOrNull(PAYMENT_DATA),
                        paymentMethodType = jsonObject.getStringOrNull(PAYMENT_METHOD_TYPE),
                    )
                } catch (e: JSONException) {
                    throw ModelSerializationException(Threeds2Action::class.java, e)
                }
            }
        }
    }

    enum class SubType(val value: String) {
        FINGERPRINT("fingerprint"),
        CHALLENGE("challenge");

        companion object {
            @JvmStatic
            fun parse(value: String): SubType {
                return when (value) {
                    FINGERPRINT.value -> FINGERPRINT
                    CHALLENGE.value -> CHALLENGE
                    else -> throw IllegalArgumentException("No Subtype matches the value of: $value")
                }
            }
        }
    }
}
